core/src/main/scala/com/raphtory/api/analysis/table/TableOutputTracker.scala

```diff
@@ -0,0 +1,135 @@
+    msg match {
+      case RowOutput(perspective, row) =>
+        resultsInProgress.getOrElseUpdate(perspective, ArrayBuffer.empty[Row]).append(row)
+      case EndPerspective(perspective) =>
+        perspectiveDoneCounts(perspective) = perspectiveDoneCounts.getOrElse(perspective, 0) + 1
+        if (perspectiveDoneCounts(perspective) == totalPartitions) {
+          perspectiveDoneCounts.remove(perspective)
+          resultsInProgress.remove(perspective).map(_.toArray) match {
+            case Some(rows) =>
+              completedResults.add(TableOutput(perspective, rows, getJobId, conf, topics))
+            case None       =>
+              completedResults.add(TableOutput(perspective, Array.empty, getJobId, conf, topics))
+          }
+        }
+      case EndOutput                   =>
+        jobsDone += 1
+        if (jobsDone == totalPartitions) {
+          completedResults.add(EndOutput)
+          stop()
+        }
+    }
+
+  override private[raphtory] def run(): Unit = {
+    logger.info(s"Job $getJobId: Starting output collector.")
+    outputListener.start()
+  }
+
+  override private[raphtory] def stop(): Unit = {
+    logger.debug(s"Stopping QueryProgressTracker for $getJobId")
+    outputListener.close()
+  }
+
+  /** Returns job identifier for the query
+    * @return job identifier
+    */
+  def getJobId: String =
+    tracker.getJobId
+
+  /** Returns the latest `Perspective` processed by the query
+    * @return latest perspective
+    */
+  def getLatestPerspectiveProcessed: Option[Perspective] =
+    tracker.getLatestPerspectiveProcessed
+
+  /** Returns list of perspectives processed for the query so far
+    * @return a list of perspectives
+    */
+  def getPerspectivesProcessed: List[Perspective] =
+    tracker.getPerspectivesProcessed
+
+  /** Returns the time taken to process each perspective in milliseconds */
+  def getPerspectiveDurations: List[Long] =
+    tracker.getPerspectiveDurations
+
+  /** Checks if job is complete
+    * @return job status
+    */
+  def isJobDone: Boolean =
+    outputDone
+
+  private def waitForNextResult(): Any =
+    if (timeout.isFinite)
+      Option(completedResults.poll(timeout.length, timeout.unit))
+    else
+      Option(completedResults.take())
+
+  override def hasNext: Boolean =
+    if (nextResult.isDefined)
+      true
+    else if (outputDone)
+      false
+    else
+      waitForNextResult() match {
+        case Some(res) =>
+          res match {
+            case EndOutput      =>
+              outputDone = true
+              false
+            case v: TableOutput =>
+              nextResult = Some(v)
+              true
+          }
+        case None      =>
+          false
+      }
+
+  override def next(): TableOutput =
+    if (hasNext) {
+      val out = nextResult.get
+      nextResult = None
+      out
+    }
+    else
+      throw new NoSuchElementException("All perspectives already processed")
```

Comment:
Change message for case where output is not done but no more results are available due to timeout